### ISSUE TITLE
[Question] Delegate deserialization to a generic `Deserialize`able in a visitor method.

### ISSUE THREAD

**Issue by tomprogrammer** [serde-rs/serde#146]
I haven't figured out how I can achieve the following using `serde`, therefore I appreciate your help.

# Introduction
The data structure I want to deserialize appears in two forms. It contains either a single string or an array of objects.
```json
["This string maps to the error variant"]
[{keys: values}, {keys: values}, …]
```

The `enum VariantData<T>` should represent this data structure where `T` is a generic `Deserialize`able encapsulating the knowlegde how to deserialize itself.
```rust
enum VariantData<T> {
    Content(Vec<T>),
    Error(String),
}
```

In `VecVisitor<T>` calling `SeqVisitor::visit()` is enough, because every element is of type `T` and the method delegates deserialization to `T`.

I can use the same approach for `VariantData<T>` deserializing a single element to a helper `enum FirstValue<T>`. I guess an alternative would be to use a `EnumVisitor` directly but I'm not yet familiar enough with it.
```rust
enum FirstValue<T> {
    Content(T),
    Error(String),
}
```

# The problem
In both cases I need to distinguish the type I want deserialize to based on the kind of structure in the raw data. If a map follows the target type is `T` (or `Vec<T>` if directly serializing the array instead of looping over it), if a plain string follows the target type is `String`.

My first idea was: (The code is incomplete since its enough to show my problem.)
```rust
impl<T> Deserialize for VariantContent<T> {
    fn deserialize<D: Deserializer>(deserializer: &mut D) -> Result<Variant<T>, D::Error> {
        struct VariantContentVisitor<T> {
            data: PhantomData<T>,
        }

        impl<T> de::Visitor for VariantContentVisitor<T> where T: Deserialize {
            type Value = VariantContent<T>;

            fn visit_seq<V>(&mut self, mut visitor: V) -> Result<Self, V::Error>
                where V: de::SeqVisitor
            {
                let first: FirstValue<T> = try!(visitor.visit());
                match first {
                    FirstValue::Content(x) => {
                        // Loop over the remaining Ts and add them to `Vec<T>`
                        let elem: T = try!(visitor.visit());
                        unimplemented!()
                    },
                    FirstValue::Error(x) => VariantContent::Error(x),
                }
            }
        }

        deserializer.visit(VariantContentVisitor { data: PhantomData })
    }
}

// …
impl<T> de::Visitor for FirstValueVisitor<T> where T: Deserialize {
    type Value = FirstValue<T>;

    fn visit_map<V>(&mut self, mut visitor: V) -> Result<Self, V::Error>
        where V: de::MapVisitor
    {
        // Deserialize a T. This is called for the first element.
        unimplemented!()
    }

    fn visit_string(&mut self, value: String) -> Result<Self, V::Error> {
        FirstValue::Error(value)
    }
}
// …
```

Now my question: How can I delegate the deserialization of a `T` to it's `Deserialize` implementation in `visit_map()`? It is a generic type, I can't know how to deserialize that type correctly. As mentioned above I also can't guide `SeqVisitor::visit` to the correct type directly, because I need to know if a map or a string follows next.

I also tried to use a `EnumVisitor`:
```rust
impl<O> Deserialize for VariantContent<O> {
    fn deserialize<D: Deserializer>(deserializer: &mut D) -> Result<Variant<O>, D::Error> {
        struct VariantContentVisitor<O> {
            data: PhantomData<O>,
        }

        impl<O> de::EnumVisitor for VariantContentVisitor<O> where O: Deserialize {
            type Value = VariantContent<O>;

            fn visit<V>(&mut self, visitor: V) -> Result<Self::Value, V::Error> {
                // How to determine which variant to decode?
            }
        }

        deserializer.visit_enum("VariantContent",
                                &["Content", "Error"],
                                VariantContentVisitor { data: PhantomData })
    }
}
```

It is shorter than my first attempt, but how can I determine which type (`T` or `String`) I need to deserialize to? It depends on the kind of the following data, map or primitive string.

Thanks for your help!

**Comment by oli-obk** [comment]
Have a look at https://github.com/serde-rs/serde/blob/master/serde_json/src/value.rs#L265 this shows how it's done for the json value enum

**Comment by tomprogrammer** [comment]
Ah, I need a visitor for the target type, the trait bound then would be `T: de::Visitor`. I can access the target type using the associated type `de::Visitor::Value`. I will try that.

Thanks

**Comment by tomprogrammer** [comment]
It seems a good idea, I introduced the type paramter `TV: de::Visitor` which can produce a value `T`. The biggest problem now seems to get a new `TV`. Maybe a constructor method should be added to the trait `de::Visitor`. I guess it's not as easy as I think, because stateful visitors would need additional parameters in their constructor method.

```rust
enum FirstValue<T> {
    Content(T),
    Error(String),
}

impl<T> Deserialize for FirstValue<T> {
    fn deserialize<D: Deserializer>(deserializer: &mut D) -> Result<Self, D::Error> {
        struct FirstValueVisitor<T> {
            data: PhantomData<T>,
        }

        impl<TV> de::Visitor for FirstValueVisitor<TV> where TV: de::Visitor {
            type Value = FirstValue<TV::Value>;

            fn visit_map<V>(&mut self, mut visitor: V) -> Result<Self::Value, V::Error>
                where V: de::MapVisitor
            {
                // Problem:
                let values = try!(TV::new().visit_map(visitor));
                unimplemented!()
            }

            fn visit_string<E>(&mut self, value: String) -> Result<Self::Value, E>
                where E: de::Error
            {
                FirstValue::Error(value)
            }
        }

        deserializer.visit(FirstValueVisitor { data: PhantomData })
    }
}
```

**Comment by oli-obk** [comment]
Whats wrong with `type Value = `

**Comment by oli-obk** [comment]
Sorry wrong button

**Comment by tomprogrammer** [comment]
Do you mean something like that:
```rust
struct FirstValueVisitor<'a, O, D: 'a> {
    data: PhantomData<O>,
    deserializer: &'a mut D,
}
```

If I put the deserializer into the visitor I can't call `deserializer.visit(visitor)` because the visitor borrows the deserializer.

**Comment by oli-obk** [comment]
No. I mean that you should call exactly like i wrote in the previous post. I don't have a computer here so i can't help you much. Sorry. Have a look at the code from serde json value, it's solving the same problem as you. Or look at the generated code from an enum deserialization

**Comment by tomprogrammer** [comment]
Now I found some time to work on this again.

My code now looks like:
```
enum FirstValue<O> {
    Content(O),
    Error(String),
}

impl<O> Deserialize for FirstValue<O> where O: Deserialize {
    fn deserialize<D: Deserializer>(deserializer: &mut D) -> Result<Self, D::Error> {
        struct FirstValueVisitor<O> {
            data: PhantomData<O>,
        }

        impl<O> de::Visitor for FirstValueVisitor<O> where O: Deserialize {
            type Value = FirstValue<O>;

            fn visit_map<V>(&mut self, mut visitor: V) -> Result<Self::Value, V::Error>
                where V: de::MapVisitor
            {
                unimplemented!();
                let values = Deserialize::deserialize(self.deserializer)
                    .or(Err(de::Error::syntax("FirstValue, visit_map")));
                values
            }

            fn visit_string<E>(&mut self, value: String) -> Result<Self::Value, E>
                where E: de::Error
            {
                Ok(FirstValue::Error(value))
            }
        }

        deserializer.visit(FirstValueVisitor { data: PhantomData })
    }
}
```

Of course I get an error:
```
src/protocol/response.rs:311:55: 311:72 error: attempted access of field `deserializer` on type `&mut protocol::response::FirstValue<O>.Deserialize::deserialize::FirstValueVisitor<O>`, but no field with that name was found
src/protocol/response.rs:311                 let values = Deserialize::deserialize(self.deserializer)
```
Because of that error I introduced a field `deserializer` in `FirstValueVisitor<O>` as I wrote in my last comment.

Maybe we talk past each other.

**Comment by tomprogrammer** [comment]
Another problem is, that at the time `Deserializer::deserialize(self.deserializer)` is called the opening brace was already consumed by `visit_map`. Therefore `deserialize` can't deserialize the object, because the next token it sees is probably a string (the key of the object). What I really want is to forward the `visit_map` call to the the visitor of the type `O`.

The reason I need this forwarding is, that `visit_map` is forwarded to a different type than `visit_string`.

**Comment by oli-obk** [comment]
>Another problem is, that at the time Deserializer::deserialize(self.deserializer) is called the opening brace was already consumed by visit_map. Therefore deserialize can't deserialize the object, because the next token it sees is probably a string (the key of the object). What I really want is to forward the visit_map call to the the visitor of the type O

you could put an `Option<u8>` field into your deserializer where you can "put back" an item you read from the stream. when reading from the stream, you always then check the field first, and if it's `None`, only then you actually read from the stream.

anyway, I think it could be much simpler. Do you have your code somewhere so I can play with it?

**Comment by tomprogrammer** [comment]
The [reduced code example](https://gist.github.com/tomprogrammer/2ffa81776b646e4b7cde) is available in a gist. I extracted it from my actual code and commented some spots, just the tests are a bit contrived.

**Comment by oli-obk** [comment]
ok, I see now... this is indeed nontrivial... (also very odd with the single element vector...)

**Comment by tomprogrammer** [comment]
Yeah, the single element vector in case of an error was not my choice, but I want to deserialize that.

**Comment by tomprogrammer** [comment]
Finally this issue is solved!

Using a `Deserializer` to forward the call is possible. In the updated [gist](https://gist.github.com/tomprogrammer/2ffa81776b646e4b7cde) you can see how I did it. In short, there is a specialized `Deserializer` which knows how to deserialize from the supplied `MapVisitor`.

Thanks @oli-obk for your patience and your tips that brought me on track and @erickt whose `de::value::*Deserializer`s gave me inspiration.

**Comment by kyegupov** [comment]
I've encountered this thread while trying to figure out how to do delegating deserialization myself; basically, I wanted to implement the logic of `if string "None" then return Self::None, else parse as T and return Self::Some(T)`.

Unfortunately, gists by @tomprogrammer are now dead, so I was stuck on a problem: ok, I've tried to deserialize as "None", and failed, how do I now deserialize as T, given that the deserializer already has been consumed?

Well, I found the answer in the depths of serde's implementation of parsing untagged enums, which also need to parse the incoming stream several times. Yes, "deserializer" is indeed supposed to be one-shot. To work around that, serde just parses everything into a JSON-like tree ("Content" struct in serde's internals), and then just tries to convert that struct to various enum options.

So it is possible to emulate this behaviour with serde_json::Value, I guess. I opted for using an untagged enum + serialize_with / deserialize_with approach.

Hope this would be helpful to someone.
